src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,314 @@
+  R[X] →+* R[T;T⁻¹] :=
+begin
+  refine ring_hom.comp _ (to_finsupp_iso R).to_ring_hom,
+  exact (map_domain_ring_hom R (nat.cast_add_monoid_hom ℤ)),
+end
+
+/--  The `R`-algebra map, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent_alg [comm_semiring R] :
+  R[X] →ₐ[R] R[T;T⁻¹] :=
+begin
+  refine alg_hom.comp _ (to_finsupp_iso_alg R).to_alg_hom,
+  exact (map_domain_alg_hom R R (nat.cast_add_monoid_hom ℤ)),
+end
+
+namespace laurent_polynomial
+
+section semiring
+variables [semiring R]
+
+lemma single_zero_one_eq_one : (single 0 1 : R[T;T⁻¹]) = (1 : R[T;T⁻¹]) := rfl
+
+/-!  ### The functions `C` and `T`. -/
+/--  The ring homomorphism `C`, including `R` into the ring of Laurent polynomials over `R` as
+the constant Laurent polynomials. -/
+def C : R →+* R[T;T⁻¹] :=
+{ to_fun    := single 0,
+  map_one'  := rfl,
+  map_mul'  := λ x y, by simp only [single_mul_single, add_zero],
+  map_zero' := single_zero,
+  map_add'  := λ x y, single_add }
+
+lemma single_eq_C (r : R) : single 0 r = C r := rfl
+
+/--  The function `n ↦ T ^ n`, implemented as a sequence `ℤ → R[T;T⁻¹]`. -/
+def T (n : ℤ) : R[T;T⁻¹] := single n 1
+
+@[simp]
+lemma T_zero : (T 0 : R[T;T⁻¹]) = 1 := rfl
+
+lemma T_add (m n : ℤ) : (T (m + n) : R[T;T⁻¹]) = T m * T n :=
+by { convert single_mul_single.symm, simp [T] }
+
+@[simp]
+lemma T_pow (m : ℤ) (n : ℕ) : (T m ^ n : R[T;T⁻¹]) = T (n * m) :=
+by rw [T, T, single_pow n, one_pow, nsmul_eq_mul, int.nat_cast_eq_coe_nat]
+
+/-- The `simp` version of `mul_assoc`, in the presence of `T`'s. -/
+@[simp]
+lemma mul_T_assoc (f : R[T;T⁻¹]) (m n : ℤ) : f * T m * T n = f * T (m + n) :=
+by simp [← T_add, mul_assoc]
+
+@[simp]
+lemma single_eq_C_mul_T (r : R) (n : ℤ) :
+  (single n r : R[T;T⁻¹]) = (C r * T n : R[T;T⁻¹]) :=
+by convert single_mul_single.symm; simp
+
+-- This lemma locks in the right changes and is what Lean proved directly.
+-- The actual `simp`-normal form of a Laurent monomial is `C a * T n`, whenever it can be reached.
+@[simp]
+lemma monomial_eq_C_mul_T (n : ℕ) (r : R) :
+  ((polynomial.monomial n r).to_laurent : R[T;T⁻¹]) = C r * T n :=
+begin
+  show map_domain (nat.cast_add_monoid_hom ℤ) ((to_finsupp_iso R) (monomial n r)) =
+    (C r * T n : R[T;T⁻¹]),
+  convert (map_domain_single : _ = single (nat.cast_add_monoid_hom ℤ n) r),
+  { exact to_finsupp_monomial n r },
+  { simp only [nat.coe_cast_add_monoid_hom, int.nat_cast_eq_coe_nat, single_eq_C_mul_T] },
+end
+
+@[simp]
+lemma C_eq_C (r : R) :
+  (polynomial.C r).to_laurent = C r :=
+begin
+  convert monomial_eq_C_mul_T 0 r,
+  simp only [int.coe_nat_zero, T_zero, mul_one],
+end
+
+@[simp]
+lemma X_eq_T :
+  (polynomial.X.to_laurent : R[T;T⁻¹]) = T 1 :=
+begin
+  have : (polynomial.X : R[X]) = monomial 1 1,
+  { simp [monomial_eq_C_mul_X] },
+  simp [this, monomial_eq_C_mul_T],
+end
+
+@[simp] lemma polynomial_one_eq_one : (polynomial.to_laurent : R[X] → R[T;T⁻¹]) 1 = 1 :=
```

Comment:
This should be called `polynomial.to_laurent_one`; but doesn't `map_one` already prove this?

Reply:
You are right: I changed the proof, but maybe you wanted me to remove the lemma altogether?